### ISSUE TITLE
Paginate Replication Jobs Async

### ISSUE THREAD

**Issue by ianseyer** [goharbor/harbor#21330]
It would appear that harbor attempts to build a list of every repository in a replication source registry. This is not scalable, as some may have a lot of repositories. This means that replication jobs sit idly, potentially allocating an array of >100,000 repositories or timing out.

However, unfortunately distribution provides no way to filter on _catalog. This means that even a very specific filter replication requires pulling a list of every single repository in a registry. Is there any work planned to make this process more scalable? I am currently trying a (sloppy) fork that runs filters per-page of results, rather than every populating a list of every single repository. However, this is still ultimately single threaded and not very efficient.

I see this as a potential solution:
- in core: grab a page of _catalog (Client.Catalog() already does pagination)
- pass that page to a worker so that it can asynchronously:
  - filter that page
  - spawn subJobs to replicate results of the filter for that page
- in core: grab the next page, repeat

This way, progress is being made inside workers rather than waiting on core for an unknown amount of time for an unknown amount of repositories to be replicated.

**Comment by wy65701436** [comment]
thanks for reporting the issue, @ianseyer. 

What timeout did you encounter during the replication process? The proposed logic change would fundamentally rework the core replication functionality. We need to first identify the root cause of the issue and explore potential enhancements or fixes based on the current design. If that doesn't resolve the problem, we can then consider modifying the core flow.

**Comment by stonezdj** [comment]
Not all replication will call the `/_catalog` API, if it is replication source is Harbor, it should call the list artifact API. the default page size (15) will slow down the stage process.

Harbor v2.12.0 already fixed this issue by change the pageSize to 100 with this pr: https://github.com/goharbor/harbor/pull/21081, you could try Harbor v2.12.0.

**Comment by cf-devtools-team** [comment]
Yes, I am using v2.12.0. A page size adjustment does not solve the issue here, as it does not run a replication job _per_ page, but instead just collects all pages into one big array. This is the phase that is not viable for large registries.

This would be for native adapter, though I see no reason why you wouldn't want replication to happen per page for all adaptors.

**Comment by wy65701436** [comment]
@ianseyer let me clarify, you're doing replication from docker-hub(or self hosted docker registry) to your harbor?

**Comment by wy65701436** [comment]
@ianseyer Did you notice that the timeout issue occurred during the stage of calling the catalog API? I believe there are two potential workarounds outside of Harbor that could help resolve this issue:

- Proxy-cache: You can configure a proxy-cache to allow all your users to pull images from Harbor, while Harbor caches images from the upstream distribution. Additionally, you could set a timeline for your organization to shut down the distribution, ensuring that all active images are available in your local Harbor instance.

- Custom Script: You can write a script to pull and push all the images from the distribution to Harbor, instead of relying on Harbor's built-in replication feature.
